lib/internal/process/esm_loader.js

```diff
@@ -32,24 +35,31 @@ exports.importModuleDynamicallyCallback = async function(wrap, specifier) {
   throw new ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING();
 };
 
-let loaderResolve;
-exports.loaderPromise = new Promise((resolve) => loaderResolve = resolve);
-
-exports.ESMLoader = undefined;
+let ESMLoader = new Loader();
+exports.ESMLoader = ESMLoader;
 
-exports.initializeLoader = function(cwd, userLoader) {
-  let ESMLoader = new Loader();
-  const loaderPromise = (async () => {
-    if (userLoader) {
-      const hooks = await ESMLoader.import(
-        userLoader, pathToFileURL(`${cwd}/`).href);
-      ESMLoader = new Loader();
-      ESMLoader.hook(hooks);
-      exports.ESMLoader = ESMLoader;
-    }
-    return ESMLoader;
+let calledInitialize = false;
+exports.initializeLoader = initializeLoader;
+function initializeLoader() {
+  assert(calledInitialize === false);
+  calledInitialize = true;
+  if (!userLoader)
+    return;
```

Comment:
Can't we just return `Promise.resolve()` here instead of doing a conditional at the only call site? That way the return value here is always a promsie.

Reply:
This is the issue with async_hooks - any promises created during bootstrap will be missed as async hooks events, but will capture later lifecycle events which creates unstable async hooks states failing lots of tests.

Reply:
Can you leave a comment here about the reasoning?

Reply:
Ah it seems I missed the full context on this comment - the suggestion you originally mentioned works here, added.